### PR TITLE
complete with the inventory system with handeling for powerups

### PR DESCRIPTION
**Pull request by TrygveJW**

### CODE REVIEW

File: unityfiles/Assets/Scripts/PlayerMovement.cs [hunk 1]
```diff
@@ -28,10 +31,15 @@ public class PlayerMovement : MonoBehaviour {
         set => jump = value;
     }
 
+    private void Start()
+    {
+        this.characterController2D = this.GetComponent<CharacterController2D>();
+    }
+
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/PlayerMovement.cs]: Brackets

**Review by perkynades — APPROVED** [review]
Was only one thing concerning brackets, but everything else is good

File: unityfiles/Assets/Scripts/Game/Enemy/EBTestMoveDown.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+using UnityEngine;
+
+/// <summary>
+/// Enemy Behavior test move down
+/// test behavior that moves the enemy downwards
+/// </summary>
+
+[RequireComponent (typeof (Enemy))]
+public class EBTestMoveDown : MonoBehaviour, IEnemyBehavior {
+
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Game/Enemy/EBTestMoveDown.cs:12]: Change the name of the class to something more readable

File: unityfiles/Assets/Scripts/Game/Enemy/Enemy.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    public EnemyType EnemyType {
+        get { return this.enemyType; }
+        set { this.enemyType = value; }
+    }
+
+    // -- public -- //
+
+
+    public void dead () {
+        this.enemyBehavior.OnDead ();
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Game/Enemy/Enemy.cs]: The first letter in method name should be upper case

File: unityfiles/Assets/Scripts/Game/Enemy/EBTestMoveDown.cs (hunk 2 above)
Comment (TrygveJW) [reply unityfiles/Assets/Scripts/Game/Enemy/EBTestMoveDown.cs:12]: EnemyBehaviorTestMoveDown er etter min mening litt langt fint om vi kann la denne gå

File: unityfiles/Assets/Scripts/Game/HealthController.cs [hunk 4]
```diff
@@ -1,52 +1,99 @@
...
+        {
+            this.isDead = false;
         }
     }
-
-    
-
-    
 }
+
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Game/HealthController.cs]: Bracket placement

File: unityfiles/Assets/Scripts/PlayerMovement.cs (hunk 1 above)
Comment (TrygveJW) [reply unityfiles/Assets/Scripts/PlayerMovement.cs]: dette er ikke mitt script ikke helt sikker på hvordan dette havnet her

File: unityfiles/Assets/Scripts/Game/Interfaces/IDamageble.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+/// <summary>
+/// Interface for any object that can receive damage
+/// </summary>
+public interface IDamageable {
+
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Game/Interfaces/IDamageble.cs]: Change class name to something more readable

File: unityfiles/Assets/Scripts/Game/Interfaces/IDamageble.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    /// <summary>
+    /// describes what to do when the character dies
+    /// </summary>
+    void dead();
+}
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Game/Interfaces/IDamageble.cs]: First letter of method should be upper case

File: unityfiles/Assets/Scripts/Game/Weapon/BulletControl.cs [hunk 7]
```diff
@@ -37,52 +38,57 @@ public float Damage
...
-        if (Time.time > this.killAt && this.isKillable){
-            this.isKillable = false;
+        if (Time.time > this.killAt && this.isActive){
+            this.isActive = false;
             this.KillSelf();
         }
     }
 
     void OnTriggerEnter2D(Collider2D Col) 
     {
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Game/Weapon/BulletControl.cs:80]: Bracket placement

File: unityfiles/Assets/Scripts/Game/Weapon/GunController.cs [hunk 8]
```diff
@@ -117,22 +143,27 @@ void Start()
...
+        if (fireRateTimer.Done("fRate", true)){
+            // if yes is there any bullets in the active bullet list that is inactive
             for (int i = this.activeBullets.Count; i > 0 ; i--)
             {
                 GameObject g = (GameObject)this.activeBullets[i-1];
-                //det her kan umulig ver veldig effektivt
+                //TODO: probably shit solution somone try somthing better
                 if (!g.activeInHierarchy){
                     //if there is move them to the inactive list
                     this.activeBullets.Remove(g);
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Game/Weapon/GunController.cs:169]: Bracket placement

File: unityfiles/Assets/Scripts/Helper/Letter_gen_Finn_pos/SudoRandomLetterGenerator.cs [hunk 9]
```diff
@@ -245,45 +227,38 @@ List<SetValuePair> GetFromReduced()
...
     {
         List<SetValuePair> possibleReductionSetts = new List<SetValuePair>();
+        string[] activeAr = this.activeLetterSet.ToArray();
         foreach (SetValuePair svp in this.allSettValuePairs.GetListFromIndex(reductionWordSize))
         {
-            if (this.IsAsubsettB(svp.letter_sett, this.activeLetterSet)){
+            if (WUArrays.IsASubsetB(svp.letter_sett, activeAr)){
                 possibleReductionSetts.Add(svp);
             }
         }
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Letter_gen_Finn_pos/SudoRandomLetterGenerator.cs]: Bracket placement

File: unityfiles/Assets/Scripts/Helper/Util/Timer.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    /// <summary>
+    /// restarts the timer
+    /// </summary>
+    public void Restart()
+    {
+        completeTime.AddMilliseconds(waitTime);
+    }
+}
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Util/Timer.cs]: Bracket placement

File: unityfiles/Assets/Scripts/Helper/Util/Timers.cs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+                }
+                retObj = true;
+            } else
+            {
+                retObj = false;
+            }
+        }
+        return retObj;
+    }
+}
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Util/Timers.cs]: change the name of "t" to something more readable

File: unityfiles/Assets/Scripts/Helper/Util/WUArrays.cs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+using System.Linq;
+
+using UnityEngine;
+
+
+static class WUArrays
+{
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Util/WUArrays.cs]: Declare if public or private, and change name of class to something more readable

File: unityfiles/Assets/Scripts/Helper/Util/WUArrays.cs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        if (bIndex <= bLen -1)
+        {
+            for (int i = bIndex; i < bLen; i++)
+            {
+                returnVals.Add(B[i]);
+            }
+        }
+
+        return returnVals;
+
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Util/WUArrays.cs]: Bracket placement

File: unityfiles/Assets/Scripts/Helper/Util/WUArrays.cs [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        // tad faster
+        LinkedList<string> tmp = new LinkedList<string>(B);
+
+        foreach (string item in A)
+        {
+            tmp.Remove(item);
+        }
+
+        List<string> rett = new List<string>(tmp);
+        return rett;
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Util/WUArrays.cs]: If this commented out code is not meant to be used later, maby consider removing it

File: unityfiles/Assets/Scripts/Helper/Util/WUGameObject.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+static class WUGameObjects{
+
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Util/WUGameObject.cs:5]: Declare if public or private, change name of the class to something more readable

File: unityfiles/Assets/Scripts/Helper/Util/WUGameObject.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            GameObject child = root.transform.GetChild(i).gameObject;
+            if (child.tag.Equals(tagLike)){
+                children.Add(child);
+            }
+        }   
+        
+        return children;
+    }
+
```
Comment (perkynades) [inline unityfiles/Assets/Scripts/Helper/Util/WUGameObject.cs:73]: Bracket placement

File: unityfiles/Assets/Scripts/Game/Spawning/Spawner.cs [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+/// <summary>
+/// Responsible for Spawning enemys.
+/// 
+/// </summary>
+public class Spawner : MonoBehaviour {
+
+    // -- editor
+    [Header ("spawning methods")]
+    [Tooltip ("spawn on and one")]
```
Comment (sanderhurlen) [inline unityfiles/Assets/Scripts/Game/Spawning/Spawner.cs:17]: one and one / en og en?

Comment (sanderhurlen) [reply unityfiles/Assets/Scripts/Game/Spawning/Spawner.cs:17]: Altså: mener du "spawn one and one?"

File: unityfiles/Assets/Scripts/Game/Spawning/Spawner.cs [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    [SerializeField]
+    private bool spawnSingle = false;
+
+    [Tooltip ("check if theere is space to spawn if not don't")]
```
Comment (sanderhurlen) [inline unityfiles/Assets/Scripts/Game/Spawning/Spawner.cs:21]: "check if there is available space to spawn"

**Review by sanderhurlen — APPROVED** [review]
Very good!

File: unityfiles/Assets/Scripts/Helper/Util/Timers.cs (hunk 11 above)
Comment (TrygveJW) [reply unityfiles/Assets/Scripts/Helper/Util/Timers.cs]: nope this is a 10 line temporary variable

File: unityfiles/Assets/Scripts/Game/Spawning/Spawner.cs (hunk 17 above)
Comment (TrygveJW) [reply unityfiles/Assets/Scripts/Game/Spawning/Spawner.cs:17]: yas english is verry god hær

**Pull request merged**